### PR TITLE
Dev/aidan/scan kt local

### PR DESCRIPTION
**Pull request by AidanBeltonS**
Moves memory into local_accessor then reduces/scans based on that. Prevents multiple reads from global memory.

wi = workitem
wg = workgroup

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 1]
```diff
@@ -163,16 +165,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+                _ONEDPL_PRAGMA_UNROLL
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i)
+                    tile_vals[local_id * elems_per_workitem + i] = __in_rng[wi_current_offset + i];
+            } else {
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i) {
+                    if (wi_current_offset + i < n)
+                        tile_vals[local_id * elems_per_workitem + i] = __in_rng[wi_current_offset + i];
+                }
+            }
+
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:191]: Missing a group_barrier here?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 2]
```diff
@@ -163,16 +165,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+            if (wg_current_offset >= n)
                 return;
+            if (wg_next_offset >= n) {
+                wg_local_memory_size = n - wg_current_offset;
+                wg_next_offset = n; // Not needed
+            }
+            auto wi_current_offset = wg_current_offset + local_id * elems_per_workitem;
+
+            // TODO: vectorize loads, where possible
+            if (wi_current_offset + elems_per_workitem < n) {
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:181]:
```suggestion
            if (wi_current_offset + elems_per_workitem <= n) {
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 3]
```diff
@@ -163,16 +166,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+            auto wg_next_offset = ((tile_id+1)*elems_in_tile);
+            size_t wg_local_memory_size = elems_in_tile;
+            if (wg_current_offset >= n)
                 return;
+            if (wg_next_offset >= n) {
+                wg_local_memory_size = n - wg_current_offset;
+                wg_next_offset = n; // Not needed
+            }
+
+            // TODO: vectorize loads, where possible
```
Comment (Alcpz) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:180]: I think this is already addressed?

**Review by Alcpz — CHANGES_REQUESTED** [review]
Minor things, otherwise LGTM

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 4]
```diff
@@ -163,16 +166,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+            if (wg_next_offset <= n) {
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:181]: Something wrong with the logic here. `wg_next_offset` is never greater than `n`, so the second branch will never be taken.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 5]
```diff
@@ -163,16 +166,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+                _ONEDPL_PRAGMA_UNROLL
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i)
+                    tile_vals[local_id + stride * i] = __in_rng[wg_current_offset + local_id + stride * i];
+            } else {
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i) {
+                    if (wg_current_offset + stride * i < n)
+                        tile_vals[local_id + stride * i] = __in_rng[wg_current_offset + stride * i];
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:188]: missing `+ local_id` here, right?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 6]
```diff
@@ -163,16 +166,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
+            }
+
+            // TODO: vectorize loads, where possible
+            if (wg_next_offset <= n) {
+                _ONEDPL_PRAGMA_UNROLL
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i)
+                    tile_vals[local_id + stride * i] = __in_rng[wg_current_offset + local_id + stride * i];
+            } else {
+                for (std::uint32_t i = 0; i < elems_per_workitem; ++i) {
+                    if (wg_current_offset + stride * i < n)
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:187]: missing `+ local_id` here too I think.

Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 5 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 4 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:178]: Ya iv fixed this by having `wg_next_offset` not set to n if too large. This means the second branch is taken in cases where we might read memory out of the allocated bounds.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 3 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 7]
```diff
@@ -163,16 +165,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
...
-            auto out_begin = __out_rng.begin() + current_offset;
-
-            if (current_offset >= n)
+            // Global load into local
+            auto wg_current_offset = (tile_id*elems_in_tile);
+            auto wg_next_offset = ((tile_id+1)*elems_in_tile);
+            size_t wg_local_memory_size = elems_in_tile;
+            if (wg_current_offset >= n)
                 return;
+            if (wg_next_offset >= n) {
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Applied suggesting

**Review by joeatodd — APPROVED** [review]
LGTM

**Review by Alcpz — APPROVED** [review]
LGTM

**Pull request merged**